File: csharp/Ice/optional/ContactDBI.cs
// Copyright (c) ZeroC, Inc.

using Demo;

#nullable enable

public class ContactDBI : ContactDBDisp_
{
    private Dictionary<string, Contact> _contacts = new Dictionary<string, Contact>();

    public override void addContact(string name, NumberType? type, string? number, int? dialGroup, Ice.Current current)
    {
        var contact = new Contact
        {
            name = name
        };
        if (type is not null)
        {
            contact.type = type;
        }
        if (number is not null)
        {
            contact.number = number;
        }
        if (dialGroup is not null)
        {
            contact.dialGroup = dialGroup;
        }
        _contacts[name] = contact;
    }

    public override void updateContact(
        string name,
        NumberType? type,
        string? number,
        int? dialGroup,
        Ice.Current current)
    {
        if (_contacts.TryGetValue(name, out Contact? c))
        {
            if (type is not null)
            {
                c.type = type;
            }
            if (number is not null)
            {
                c.number = number;
            }
            if (dialGroup is not null)
            {
                c.dialGroup = dialGroup;
            }
        }
    }

    public override Contact? query(string name, Ice.Current current)
    {
        if (_contacts.TryGetValue(name, out Contact? c))
        {
            return c;
        }
        return null;
    }

    public override string? queryNumber(string name, Ice.Current current)
    {
        if (_contacts.TryGetValue(name, out Contact? c))
        {
            return c.number;
        }
        return null;
    }

    public override void queryDialgroup(string name, out int? dialGroup, Ice.Current current)
    {
        if (_contacts.TryGetValue(name, out Contact? c))
        {
            dialGroup = c.dialGroup;
        }
        else
        {
            dialGroup = null;
        }
    }

    public override void shutdown(Ice.Current current)
    {
        System.Console.Out.WriteLine("Shutting down...");
        current.adapter.getCommunicator().shutdown();
    }
}

File: cpp/Ice/optional/ContactDBI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include "ContactDBI.h"
#include <Ice/Ice.h>
#include <iostream>

using namespace std;
using namespace Demo;

void
ContactDBI::addContact(
    string name,
    optional<NumberType> type,
    optional<string> number,
    optional<int> dialGroup,
    const Ice::Current&)
{
    auto contact = make_shared<Contact>();
    contact->name = name;
    if (type)
    {
        contact->type = std::move(type);
    }
    if (number)
    {
        contact->number = std::move(number);
    }
    if (dialGroup)
    {
        contact->dialGroup = std::move(dialGroup);
    }
    auto p = _contacts.insert(make_pair(name, contact));
    if (!p.second)
    {
        p.first->second = contact;
    }
}

void
ContactDBI::updateContact(
    string name,
    optional<NumberType> type,
    optional<string> number,
    optional<int> dialGroup,
    const Ice::Current&)
{
    auto p = _contacts.find(name);
    if (p != _contacts.end())
    {
        if (type)
        {
            p->second->type = type;
        }
        if (number)
        {
            p->second->number = number;
        }
        if (dialGroup)
        {
            p->second->dialGroup = dialGroup;
        }
    }
}

shared_ptr<Contact>
ContactDBI::query(string name, const Ice::Current&)
{
    auto p = _contacts.find(name);
    if (p != _contacts.end())
    {
        return p->second;
    }
    return nullptr;
}

optional<string>
ContactDBI::queryNumber(string name, const Ice::Current&)
{
    auto p = _contacts.find(name);
    if (p != _contacts.end())
    {
        return p->second->number;
    }
    return nullopt;
}

void
ContactDBI::queryDialgroup(string name, optional<int>& dialGroup, const Ice::Current&)
{
    auto p = _contacts.find(name);
    if (p != _contacts.end())
    {
        dialGroup = p->second->dialGroup;
    }
}

void
ContactDBI::shutdown(const Ice::Current& c)
{
    cout << "Shutting down..." << endl;
    c.adapter->getCommunicator()->shutdown();
}
